### ISSUE TITLE
Multi-touch and Pen do not work on 4.14.x

### ISSUE THREAD

**Issue by cydia2020** [jakeday/linux-surface#8]
Hey mate

Pen does not seem to work on my Surface Book (not performance base)

Is there a fix to this?

Cheers

`sudo lsinput` result
/dev/input/event0
   bustype : BUS_HOST
   vendor  : 0x0
   product : 0x5
   version : 0
   name    : "Lid Switch"
   phys    : "PNP0C0D/button/input0"
   bits ev : EV_SYN EV_SW

/dev/input/event1
   bustype : BUS_HOST
   vendor  : 0x0
   product : 0x6
   version : 0
   name    : "Video Bus"
   phys    : "LNXVIDEO/video/input0"
   bits ev : EV_SYN EV_KEY

/dev/input/event2
   bustype : BUS_HOST
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "Surface Pro 3/4 Buttons"
   phys    : "MSHW0040/buttons"
   bits ev : EV_SYN EV_KEY

/dev/input/event3
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard Keybo"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_MSC EV_LED EV_REP

/dev/input/event4
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard Mouse"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_REL EV_MSC

/dev/input/event5
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard Consu"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_REL EV_ABS EV_MSC

/dev/input/event6
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event7
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard Touch"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_ABS

/dev/input/event8
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event9
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event10
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event11
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event12
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event13
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event14
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event15
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event16
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event17
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event18
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event19
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event20
   bustype : BUS_USB
   vendor  : 0x45e
   product : 0x7cd
   version : 273
   name    : "Microsoft Surface Keyboard UNKNO"
   phys    : "usb-0000:00:14.0-1.4/input0"
   uniq    : ""
   bits ev : EV_SYN EV_ABS

/dev/input/event21
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH Mic"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event22
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH Headphone"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event23
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH HDMI/DP,pcm=3"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event24
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH HDMI/DP,pcm=7"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event25
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH HDMI/DP,pcm=8"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event26
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH HDMI/DP,pcm=9"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event27
   bustype : (null)
   vendor  : 0x0
   product : 0x0
   version : 0
   name    : "HDA Intel PCH HDMI/DP,pcm=10"
   phys    : "ALSA"
   bits ev : EV_SYN EV_SW

/dev/input/event28
   bustype : (null)
   vendor  : 0x1b96
   product : 0x5e
   version : 0
   name    : "ipts 1B96:005E UNKNOWN"
   phys    : "heci3"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_ABS EV_MSC

/dev/input/event29
   bustype : (null)
   vendor  : 0x1b96
   product : 0x5e
   version : 0
   name    : "ipts 1B96:005E Pen"
   phys    : "heci3"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_ABS EV_MSC

/dev/input/event30
   bustype : (null)
   vendor  : 0x1b96
   product : 0x5e
   version : 0
   name    : "ipts 1B96:005E"
   phys    : "heci3"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_ABS

/dev/input/event31
   bustype : (null)
   vendor  : 0x1b96
   product : 0x5e
   version : 0
   name    : "ipts 1B96:005E Mouse"
   phys    : "heci3"
   uniq    : ""
   bits ev : EV_SYN EV_KEY EV_REL EV_MSC

**Comment by jakeday** [comment]
According to you lsinput, the pen is detected and setup just fine. Did you pair and enable the correct mode?

**Comment by xabotage** [comment]
Love the work you've done here, thank you for making this so simple. 

Just installed 4.13, experiencing the same issue. From reddit (https://www.reddit.com/r/SurfaceLinux/comments/6tbnqx/jakedays_kernel_for_surface_bookpro_4/dmfd1p8/), it appears that this is because multitouch is not enabled. However, as the commenter notes, attempting to enable multitouch via `# echo 1 > /sys/kernel/debug/ipts/mode` causes the module to crash.

You asked if the pen is paired and if the "correct mode" is enabled. Does this mean enabling "multitouch" mode, as the reddit commenter notes?

It would be good to point out to newbies like me that pairing the pen requires that you hold down the eraser button for 7 seconds until the light flashes (I'm such an idiot I was stuck for half an hour until I looked it up on a Microsoft forum). But even more importantly, A quick blurb for how to change to multitouch mode will be super helpful once the kernel module error is fixed (is there a simpler way than executing the above command as root?)

**Comment by jakeday** [comment]
That's probably because that's not opened up in my kernels due to changes in 4.12+ of the mainline kernels. You'll have to change the default and the kernel and rebuild. Maybe I'll add it add in one of my next builds.

As for pairing the pen, yeah that's probably a good idea. However, I was just intending to build a kernel with full support more so than give a how to guide on using the Surface lol! :)

**Comment by Shadoukun** [comment]
Is changing the default in the kernel easy? Would it be a bother to write instructions on doing that until you have time to fix the debugfs?

And thanks again, your kernel is awesome.

**Comment by auspex-nl** [comment]
I tried to set the default, but i'm unsure where to change this, I changed this line (144) in `kernel/drivers/misc/ipts/ipts-msg-handler.c`:

```
ipts->sensor_mode = TOUCH_SENSOR_MODE_HID; /* start with HID */
```
to:
```
ipts->sensor_mode = TOUCH_SENSOR_MODE_RAW_DATA; /* start with RAW_DATA */
```

Recompiled the kernel, but now the module crashes during boot with same error messages as before (null pointer dereference). Not sure if this is the right place to change the default or that it will still start in the default mode and then trigger a switch and crash. Didn't have much time to look into the code any further

**Comment by xabotage** [comment]
I tried to adding:
 `ipts_switch_sensor_mode(ipts, TOUCH_SENSOR_MODE_RAW_DATA);` 
to `ipts_start()` in `kernel/drivers/misc/ipts/ipts-msg-handler.c` after it already initializes the sensor mode (right after the line `ret = ipts_handle_cmd(ipts, TOUCH_SENSOR_NOTIFY_DEV_READY_CMD, NULL, 0);`, but that didn't work either (same null pointer error). But then, I don't really know what I'm doing.

Honestly, everything works in the stock 4.10.0-35 Ubuntu 17.04 kernel except pen/touchscreen, so I tend to boot only from that one anyway since I need Virtualbox (doesn't work in the custom 4.13 kernel, which might be fixed with more fiddling that I don't have time for). I'm no kernel dev, so if I really want that pen working, it looks like the only choice is SuperKernelTouch+ for now

**Comment by jakeday** [comment]
I'm rewriting this in my 4.14.x series kernel. I'm not finished, but it may work now!

**Comment by geoffholden** [comment]
Still not working in -rc6
```
# cat /sys/kernel/debug/ipts/status
sensor mode : 0        
ipts state : 3
# echo 1 > /sys/kernel/debug/ipts/mode
# cat /sys/kernel/debug/ipts/status                                                    
sensor mode : 1        
ipts state : 1
```

Kernel output:
```
[  270.864833] ipts mei::3e8d0870-271a-4208-8eb5-9acb9402ae04:0F: cmd : 0x80000005, status 9                                                                                                   
[  270.867776] BUG: unable to handle kernel NULL pointer dereference at 0000000000000028                                                                                                       
[  270.867954] IP: intel_ipts_connect+0x19/0x90 [i915]                                                                                                                                         
[  270.867963] PGD 0 P4D 0                                                                                                                                                                     
[  270.867980] Oops: 0000 [#1] SMP                                                                                                                                                             
[  270.867989] Modules linked in: uhid algif_hash algif_skcipher af_alg rfcomm cmac bnep btusb btrtl btbcm btintel bluetooth ecdh_generic intel_ipts 8250_dw nls_iso8859_1 snd_hda_codec_hdmi intel_rapl x86_pkg_temp_thermal snd_soc_skl intel_powerclamp snd_soc_skl_ipc coretemp kvm_intel kvm snd_hda_codec_realtek irqbypass snd_hda_codec_generic snd_soc_sst_ipc snd_soc_sst_dsp crct10dif_pclmul snd_hda_ext_core snd_soc_sst_match crc32_pclmul ghash_clmulni_intel snd_soc_core pcbc snd_compress ac97_bus snd_pcm_dmaengine snd_hda_intel snd_hda_codec snd_hda_core snd_hwdep snd_pcm aesni_intel aes_x86_64 crypto_simd glue_helper cryptd intel_cstate intel_rapl_perf snd_seq_midi snd_seq_midi_event snd_rawmidi snd_seq snd_seq_device snd_timer snd mwifiex_pcie mwifiex cfg80211 soundcore mei_me shpchp mei cdc_ether                                                                                                                                                  
[  270.868172]  usbnet mii hid_multitouch input_leds joydev hid_sensor_rotation hid_sensor_als hid_sensor_gyro_3d hid_sensor_accel_3d idma64 hid_sensor_trigger virt_dma industrialio_triggered_buffer kfifo_buf hid_sensor_iio_common ov5693(C) industrialio processor_thermal_device intel_lpss_pci v4l2_common intel_soc_dts_iosf intel_pch_thermal videodev media mac_hid intel_lpss_acpi intel_lpss acpi_pad surfacepro3_button soc_button_array int3400_thermal tpm_crb dptf_power int3403_thermal acpi_thermal_rel int340x_thermal_zone parport_pc ppdev lp parport ip_tables x_tables autofs4 hid_generic hid_microsoft usbhid uas usb_storage hid_sensor_hub intel_ishtp_hid hid i915 i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt nvme fb_sys_fops nvme_core drm intel_ish_ipc intel_ishtp wmi pinctrl_sunrisepoint                                                                                                                                             
[  270.868344]  video pinctrl_intel                                                                                                                                                            
[  270.868363] CPU: 1 PID: 569 Comm: ipts_event_thre Tainted: G         C      4.14.0-rc6-custom-surface #7                                                                                    
[  270.868369] Hardware name: Microsoft Corporation Surface Pro/Surface Pro, BIOS 231.1662.769 03/30/2017                                                                                      
[  270.868377] task: ffff935edccd5e00 task.stack: ffffa179c1290000                                                                                                                             
[  270.868521] RIP: 0010:intel_ipts_connect+0x19/0x90 [i915]                                                                                                                                   
[  270.868530] RSP: 0018:ffffa179c1293c28 EFLAGS: 00010246                                                                                                                                     
[  270.868539] RAX: 0000000000000000 RBX: ffff935edcc88018 RCX: 0000000000000010                                                                                                               
[  270.868546] RDX: 0000000000000014 RSI: ffffa179c1293e90 RDI: ffffa179c1293c40                                                                                                               
[  270.868553] RBP: ffffa179c1293c30 R08: 000000000001f4e0 R09: ffffffffc06347dc                                                                                                               
[  270.868559] R10: ffffa179c1293cc8 R11: 000000000000026a R12: ffff935edcc88018                                                                                                               
[  270.868567] R13: 0000000080000007 R14: ffff935edcc88018 R15: ffff935edccd5e00                                                                                                               
[  270.868576] FS:  0000000000000000(0000) GS:ffff935eef480000(0000) knlGS:0000000000000000                                                                                                    
[  270.868585] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033                                                                                                                               
[  270.868593] CR2: 0000000000000028 CR3: 000000010d409003 CR4: 00000000003606e0                                                                                                               
[  270.868601] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000                                                                                                               
[  270.868609] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400                                                                                                               
[  270.868616] Call Trace:                                                                                                                                                                     
[  270.868641]  ipts_open_gpu+0x55/0x130 [intel_ipts]                                                                                                                                          
[  270.868657]  ? ipts_dbg_thread+0x110/0x110 [intel_ipts]                                                                                                                                     
[  270.868670]  ? ipts_set_output_buffer+0x30/0x30 [intel_ipts]                                                                                                                                
[  270.868685]  ipts_init_kernels+0x2f/0x100 [intel_ipts]                                                                                                                                      
[  270.868693]  ? ipts_init_kernels+0x2f/0x100 [intel_ipts]                                                                                                                                    
[  270.868703]  ipts_allocate_raw_data_resource+0x12/0x20 [intel_ipts]                                                                                                                         [  270.868713]  ipts_handle_resp+0x403/0x5b0 [intel_ipts]                                                                                                                                      
[  270.868727]  ? find_next_bit+0xb/0x10                                                                                                                                                       
[  270.868739]  ? __switch_to+0x1f3/0x450                                                                                                                                                      
[  270.868751]  ? __schedule+0x3ca/0x890                                                                                                                                                       
[  270.868771]  ? mei_io_cb_free+0x4c/0x50 [mei]                                                                                                                                               
[  270.868787]  ? __mei_cl_recv+0x67/0x1f0 [mei]                                                                                                                                               
[  270.868798]  ? wait_woken+0x80/0x80                                                                                                                                                         
[  270.868808]  ipts_mei_cl_event_thread+0x52/0xc0 [intel_ipts]                                                                                                                                
[  270.868822]  kthread+0x109/0x140                                                                                                                                                            
[  270.868830]  ? init_work_func+0x30/0x30 [intel_ipts]                                                                                                                                        
[  270.868840]  ? kthread_create_on_node+0x70/0x70                                                                                                                                             
[  270.868850]  ? do_syscall_64+0x61/0x120                                                                                                                                                     
[  270.868861]  ? SyS_exit_group+0x14/0x20
[  270.868869]  ret_from_fork+0x25/0x30                                                                                                                                                        
[  270.868876] Code: e8 0d 84 4a fe e9 2a fd ff ff e8 43 e3 2f fe 0f 1f 00 0f 1f 44 00 00 80 3d 66 60 08 00 00 48 8b 05 6d 5f 08 00 55 48 89 e5 41 52 <48> 8b 50 28 74 67 48 85 ff b8 ea ff ff ff 74 59 83 7f 18 01 77                                                                                                                                                                        
[  270.869121] RIP: intel_ipts_connect+0x19/0x90 [i915] RSP: ffffa179c1293c28                                                                                                                  
[  270.869126] CR2: 0000000000000028                                                                                                                                                           
[  270.869134] ---[ end trace 27af9f5092ba90ad ]---
```

**Comment by cydia2020** [comment]
Yep, the kernel module still crashes of mode is set to 1, tested on Kali, tried both self compiled and prebuilt kernels
________________________________

**Comment by jakeday** [comment]
Can you try testing the pen without adjusting the mode?

**Comment by cydia2020** [comment]
I did and it does not work unfortunately.
________________________________

**Comment by peter-sa** [comment]
I believe that I have some more patches that make the pen/multitouch mode work.  I will make a pull request with them in the next day or two.

**Comment by peter-sa** [comment]
I have pushed some patches to https://github.com/peter-sa/linux-sb/commits/for-jakeday.  I believe that these should fix several of the issues that occur with the current kernel (including the null pointer dereference) by accounting for various recent DRM changes, but have not yet had a chance to test this particular kernel on my own hardware.

**Comment by jakeday** [comment]
@peter-sa As mentioned, the patches you pushed do not impact multitouch or the pen. They are similar changes to what I had in 4.13 and cause some odd behavior in 4.14.

**Comment by geoffholden** [comment]
> @peter-sa As mentioned, the patches you pushed do not impact multitouch or the pen. They are similar changes to what I had in 4.13 and cause some odd behavior in 4.14.
I can confirm that these patches do not fix the null pointer dereference.

**Comment by peter-sa** [comment]
@jakeday  @geoffholden  I am sorry to hear that these changes do not appear to fix the behavior.  Are there any changes at all in the log?  On the kernel that I run, I had similar null pointer errors which were solved by these changes.  I will try again to compare the source trees between my kernel and yours.

**Comment by geoffholden** [comment]
The backtrace is slightly different.
```[  136.366437] Call Trace:                     
[  136.366465]  ipts_open_gpu+0x55/0x130 [intel_ipts]                                          
[  136.366481]  ? ipts_dbg_thread+0x110/0x110 [intel_ipts]                                     
[  136.366494]  ? ipts_set_output_buffer+0x30/0x30 [intel_ipts]                                
[  136.366512]  ? update_load_avg+0x429/0x590  
[  136.366525]  ? update_curr+0xf5/0x1b0       
[  136.366541]  ipts_init_kernels+0x2f/0xf0 [intel_ipts]                                       
[  136.366554]  ? ipts_init_kernels+0x2f/0xf0 [intel_ipts]                                     
[  136.366569]  ipts_allocate_raw_data_resource+0x12/0x20 [intel_ipts]                         
[  136.366582]  ipts_handle_resp+0xa4/0x5f0 [intel_ipts]                                       
[  136.366598]  ? __switch_to+0x1f3/0x460      
[  136.366611]  ? put_prev_entity+0x23/0xf0    
[  136.366637]  ? mei_io_cb_free.part.11+0x43/0x50 [mei]                                       
[  136.366651]  ? kfree+0x14a/0x160            
[  136.366662]  ? kfree+0x14a/0x160            
[  136.366683]  ? mei_io_cb_free.part.11+0x43/0x50 [mei]                                       
[  136.366704]  ? mei_io_cb_free+0x13/0x20 [mei]                                               
[  136.366727]  ? __mei_cl_recv+0x6a/0x200 [mei]                                               
[  136.366739]  ? wait_woken+0x80/0x80         
[  136.366755]  ipts_mei_cl_event_thread+0x52/0xc0 [intel_ipts]                                
[  136.366774]  kthread+0x125/0x140            
[  136.366787]  ? init_work_func+0x30/0x30 [intel_ipts]                                        
[  136.366800]  ? kthread_create_on_node+0x70/0x70                                             
[  136.366815]  ret_from_fork+0x25/0x30```

**Comment by jakeday** [comment]
@peter-sa It's probably fine, I already have these things addressed in 4.14 and I'm just waiting for the final release. They still won't fix multitouch or pen.

**Comment by peter-sa** [comment]
@geoffholden thanks for the backtrace.  Are the other parts of the OOPS entry the same?

@jakeday Interesting. Apologies for the false hope.  I can confirm that I have multitouch and pen working on a 4.14 kernel, so I will again try see if there are any differences between my kernel and this one.  I am very surprised though that these do not improve multitouch/pen because they both make changes in the multitouch-related (gfx - SingleTouch/HID does not use GuC) code that make it compliant with recent DRM changes.

**Comment by peter-sa** [comment]
I have found one other important change, although I do not know if it will fix the issue.  It is part of my doorbell re-work so I lost it.  In `i915_guc_ipts_reacquire_doorbell`, the line `int err = __guc_allocate_doorbell(guc, guc->ipts_client)` should be `int err = __guc_allocate_doorbell(guc, guc->ipts_client->stage_id)`.  @jakeday apologies if I have rediscovered an unpublished change of yours again.

**Comment by jakeday** [comment]
@peter-sa No worries! We'll get there! I just don't want to introduce new errors or weird behavior just to turn on multitouch if it's not stable yet.

**Comment by peter-sa** [comment]
Also @jakeday on rc7 do you have a "Unexpected gfp: 0x4 (GFP_DMA32)" warning? It appears to be coming from ipts_allocate_default_resource() which is not multitouch/pen/raw_data related and so should be present even with single-touch.

**Comment by jakeday** [comment]
@peter-sa I had that on rc7 at first, but I have it fixed in preparation for rc8.

**Comment by jakeday** [comment]
@peter-sa @geoffholden I'll push all my changes from 4.13 and what I'm holding from 4.14 tonight so you can go ahead and look and test.

**Comment by peter-sa** [comment]
> @peter-sa @geoffholden I'll push all my changes from 4.13 and what I'm holding from 4.14 tonight so you can go ahead and look and test.
Great, that should help with debugging, thanks!
> @peter-sa @geoffholden I'll push all my changes from 4.13 and what I'm holding from 4.14 tonight so you can go ahead and look and test.
Does that refer to the unexpected gfp, or the stage id passing in __guc_allocate_doorbell?

**Comment by peter-sa** [comment]
@geoffholden Does anything change if you boot without the ipts debug mode set and/or change the debug mode default state?  I have been booting without debug mode, so it is possible that there is some race where resources must be initialized early in the boot process but are not if you switch later.

Also, what hardware are you on?  It is perhaps possible that SP4 is different from SB or some such.

**Comment by peter-sa** [comment]
@jakeday @geoffholden  I cannot reproduce the issue at all.  I have built a kernel which is stock 4.7 + the changes in drivers/gpu/drm/i915 and drivers/misc/ipts from the for-jakeday branch of my repo (including booting in debug mode and switching later).  Although suspend/resume still causes catastrophic failure because of the GuC doorbell being lost, if I switch the mode immediately after boot I have no issue---no null pointer dereference, and libinput debug-gui shows full 10-point touch.

**Comment by geoffholden** [comment]
I can possibly check booting without debug mode set tomorrow, if you tell
me what I have to do to accomplish that.

As for hardware, I'm on a 2017 Surface Pro

**Comment by peter-sa** [comment]
@geoffholden Switching out of debug mode requires commenting out/removing the `#define ENABLE_IPTS_DEBUG` in `drivers/misc/ipts/ipts.h`.  Unfortunately, it seems unlikely that that will fix the issue, as I still have no problem even with debug mode.

I don't have a 2017 Surface Pro around that I can test with unfortunately.  It is possible that there are some major hardware/firmware differences that cause issues.  Could you try booting with something like `drm.debug=0x1e log_buf_len=4M` and send me the logs?  Those might be helpful, although I am not sure.

**Comment by jakeday** [comment]
@peter-sa Could you point out the location of your source for your 4.14.0 kernel with working multitouch and pen? I'd like to take a look!

**Comment by geoffholden** [comment]
@peter-sa I've booted with the DRM debugging enabled, the output from dmesg is here: https://gist.github.com/geoffholden/8a676b097a77074e7887ffdc775ba5c5

**Comment by jakibaki** [comment]
Could anyone who still has the 4.13 build please upload it here? The google-drive link from the 4.13 branch unfortunately points to the folder where only 4.14 is.

**Comment by jakeday** [comment]
What are you wanting 4.13 for? It was never finished.

**Comment by jakibaki** [comment]
What issues (other than the camera) were there? I want it because of the broken multitouch support in 4.14.

**Comment by jakeday** [comment]
I never finished porting IPTS to 4.13 and went straight to 4.14. You're not going to get multi-touch on 4.13 either. Multi-touch will be working shortly on 4.14, I'm just waiting on the official release.

**Comment by jakibaki** [comment]
Ah, ok. I was confused by the fact that in the 4.13 branch both touch and pen were listed as working.

Thanks for the information!

**Comment by jakeday** [comment]
That's my fault. I put work into the 4.13 build so multi-touch would be supported, but never finished the build anyway. Sorry for not being clear!

**Comment by jakeday** [comment]
I re-uploaded the 4.14.0-rc7 builds (now named jakeday42 instead of custom). This adds better support for multi-touch and pen (meaning won't completely crash the system) but still isn't 100% due to changes in the i915 modules. I'm planning on finishing these after rc8 drops.

**Comment by jakeday** [comment]
I'll push the source changes later tonight, still need to look into a resp msg handling issue. The latest build fixes the null pointer dereference error though.

**Comment by navneet92** [comment]
@jakeday I have tried to use your new 4.14.0-rc7 build. But, it not booting. My surface is getting stuck during boot.

**Comment by jakeday** [comment]
Did the previous 4.14.0-rc7 build boot for you?

**Comment by geoffholden** [comment]
The new build boots fine here. As you said, the null pointer dereference is gone, but there's still a resp msg handling issue stopping multitouch from working.

**Comment by jakeday** [comment]
Yep, just like I said. :)

**Comment by navneet92** [comment]
yes..gt it. Some configuration issues due to multiple kernels. But, still multitouch problem.

**Comment by jakeday** [comment]
Yep, multitouch is not complete for 4.14.0 yet. I'm working on it for the next rc (rc8).

**Comment by jakeday** [comment]
Finished porting my changes over from 4.13 to 4.14. You can switch between modes without error or crashing now, and ipts is correctly loaded with the gpu for multi-touch mode. Still some things to work out as I mentioned with the i915 changes in 4.14.x, and those should be mostly settled in rc8 this Sunday. I'll push and upload here shortly.

**Comment by jakeday** [comment]
The changes have been pushed to github and the prebuilt image files have been uploaded.

**Comment by jakeday** [comment]
Just to note, multi touch and pen both now work if you either a) start out in multi touch mode or b) re-initialize your input devices after switching modes. I'll make this automatic for the official release of 4.14.
